climada/hazard/tc_tracks.py

```diff
@@ -1587,17 +1683,88 @@ def ibtracs_track_agency(ds_sel):
     track_agency_ix : xarray.DataArray of ints
         For each entry in `ds_sel`, the agency to use, given as an index into `agency_pref`.
     """
-    agency_pref = IBTRACS_AGENCIES.copy()
+    agency_pref = ["wmo"] + IBTRACS_AGENCIES
     agency_map = {a.encode('utf-8'): i for i, a in enumerate(agency_pref)}
     agency_map.update({
         a.encode('utf-8'): agency_map[b'usa'] for a in IBTRACS_USA_AGENCIES
     })
     agency_map[b''] = agency_map[b'wmo']
     agency_fun = lambda x: agency_map[x]
-    track_agency = ds_sel.wmo_agency.where(ds_sel.wmo_agency != '', ds_sel.usa_agency)
-    track_agency_ix = xr.apply_ufunc(agency_fun, track_agency, vectorize=True)
+    if "track_agency" not in ds_sel.data_vars.keys():
+        ds_sel['track_agency'] = ds_sel.wmo_agency.where(ds_sel.wmo_agency != b'',
+                                                         ds_sel.usa_agency)
+    track_agency_ix = xr.apply_ufunc(agency_fun, ds_sel.track_agency, vectorize=True)
     return agency_pref, track_agency_ix
 
+def ibtracs_add_official_variable(ibtracs_ds, tc_var, add_3h=False):
+    """Add variables for the officially responsible agencies to an IBTrACS dataset
+
+    This function adds new variables to the xarray.Dataset `ibtracs_ds` that contain values of the
+    specified TC variable `var` that have been reported by the officially responsible agencies.
+    For example, if `tc_var` is "wind", there will be a new variable "official_wind" and, if
+    `add_3h` is True, an additional variable "official_3h_wind".
+
+    Parameters
+    ----------
+    ibtracs_ds : xarray.Dataset
+        Subselection of original IBTrACS NetCDF dataset.
+    tc_var : str
+        Name of variable for which to add an "official" version, e.g. "lat", "wind", "pres".
+    add_3h : bool, optional
+        Optionally, add an "official_3h" version where also 3-hourly data by the officially
+        reporting agencies is included (if available). Default: False
+    """
+    if "nan_var" not in ibtracs_ds.data_vars.keys():
+        # add an array full of NaN as a fallback value in the procedure
+        ibtracs_ds['nan_var'] = xr.full_like(ibtracs_ds.lat, np.nan)
+
+    # determine which of the official agencies report this variable at all
+    available_agencies = [a for a in IBTRACS_AGENCIES
+                          if f'{a}_{tc_var}' in ibtracs_ds.data_vars.keys()]
+    agency_map = {a.encode("utf-8"): i + 1 for i, a in enumerate(available_agencies)}
+    agency_map.update({
+        a.encode('utf-8'): agency_map[b'usa'] for a in IBTRACS_USA_AGENCIES
+    })
+
+    # map all non-reporting agency variables to the 'nan_var'
+    agency_map[b''] = 0
```

Comment:
Another curiosity: is there a good reason to use byte-literals in this case?

Reply:
For each storm and each time step the IBTrACS dataset contains information about the officially responsible agency. Byte-literals are used for this purpose in IBTrACS. If no agency is responsible for a particular time step (because it's an unofficial time-step), this field is set to `b''`.

Reply:
Thanks!